nixos/modules/services/mail/public-inbox.nix

```diff
@@ -0,0 +1,517 @@
+{ lib, pkgs, config, ... }:
+
+with lib;
+
+let
+  cfg = config.services.public-inbox;
+  stateDir = "/var/lib/public-inbox";
+
+  manref = name: vol: "<citerefentry><refentrytitle>${name}</refentrytitle><manvolnum>${toString vol}</manvolnum></citerefentry>";
+
+  gitIni = pkgs.formats.gitIni { listsAsDuplicateKeys = true; };
+  iniAtom = gitIni.type.functor.wrapped.functor.wrapped;
+
+  useSpamAssassin = cfg.settings.publicinboxmda.spamcheck == "spamc" ||
+                    cfg.settings.publicinboxwatch.spamcheck == "spamc";
+
+  publicInboxDaemonOptions = proto: defaultPort: {
+    args = mkOption {
+      type = with types; listOf str;
+      default = [];
+      description = "Command-line arguments to pass to ${manref "public-inbox-${proto}d" 1}.";
+    };
+    port = mkOption {
+      type = with types; nullOr (either str port);
+      default = defaultPort;
+      description = ''
+        Listening port.
+        Beware that public-inbox uses well-known ports number to decide whether to enable TLS or not.
+        Set to null and use <code>systemd.sockets.public-inbox-${proto}d.listenStreams</code>
+        if you need a more advanced listening.
+      '';
+    };
+    cert = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "/path/to/fullchain.pem";
+      description = "Path to TLS certificate to use for connections to ${manref "public-inbox-${proto}d" 1}.";
+    };
+    key = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "/path/to/key.pem";
+      description = "Path to TLS key to use for connections to ${manref "public-inbox-${proto}d" 1}.";
+    };
+  };
+
+  serviceConfig = srv:
+    let proto = removeSuffix "d" srv;
+        needNetwork = builtins.hasAttr proto cfg && cfg.${proto}.port == null;
+    in {
+    # Enable JIT-compiled C (via Inline::C)
+    Environment = [ "PERL_INLINE_DIRECTORY=/run/public-inbox-${srv}/perl-inline" ];
+    # NonBlocking is REQUIRED to avoid a race condition
+    # if running simultaneous services.
+    NonBlocking = true;
+    #LimitNOFILE = 30000;
+    User = config.users.users."public-inbox".name;
+    Group = config.users.groups."public-inbox".name;
+    RuntimeDirectory = [
+      "public-inbox-${srv}/perl-inline"
+      # Create RootDirectory= in the host's mount namespace.
+      "public-inbox-${srv}/root"
+    ];
+    RuntimeDirectoryMode = "700";
+    # Avoid mounting RootDirectory= in the own RootDirectory= of ExecStart='s mount namespace.
+    InaccessiblePaths = ["-+/run/public-inbox-${srv}/root"];
+    # This is for BindPaths= and BindReadOnlyPaths=
+    # to allow traversal of directories they create in RootDirectory=.
+    UMask = "0066";
+    RootDirectory = "/run/public-inbox-${srv}/root";
+    RootDirectoryStartOnly = true;
+    WorkingDirectory = stateDir;
+    MountAPIVFS = true;
+    BindReadOnlyPaths = [
+      builtins.storeDir
```

Comment:
Nit: Have you checked out the `systemd.services.*.confinement.*` options? They allow you to sandbox a service to only accessing the store paths it actually depends on, instead of anything in the store.

Reply:
@Gaelan, indeed, since this module is not using `DynamicUser` nor `RootDirectoryStartOnly`, `systemd-confinement` should be doable.
Of course `ProtectSystem` will have to be disabled (it would override the `tmpfs` built by `systemd-confinement`). As a side note, `systemd-confinement` has [an assert for that](https://github.com/NixOS/nixpkgs/blob/b99a500a045e73fcabfb141b33fe0b9021966040/nixos/modules/security/systemd-confinement.nix#L163) but in terms of usability this should probably just be a `mkForce false`.

However `systemd-confinement`'s profiles are [hard to write and maintain](http://cgit.notk.org/asmadeus/nixos-config.git/tree/profiles/postgresql.nix#n14%C2%B7), so I'm not really up for that task.

Best would be to start by writing `AppArmor` profiles :
1. they can include usual rules (like `abstractions/base`, `abstractions/nameservice`, etc.);
2. `apparmorRulesFromClosure` can handle store paths as done by `systemd-confinement`;
3. the `complain` mode produce a log helping a lot to generate profiles using `aa-logprof`.

Both `systemd-confinement` and `AppArmor` can be used together, once written `systemd-confinement` is enabled by default which is not the case of `AppArmor` (which is behind `security.apparmor.enable`), but `AppArmor` profiles is easier to start with.